artiq/gateware/pdq/core.py

```diff
@@ -0,0 +1,227 @@
+# Copyright 2013-2017 Robert Jordens <jordens@gmail.com>
+#
+# This file is part of pdq.
+#
+# pdq is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# pdq is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with pdq.  If not, see <http://www.gnu.org/licenses/>.
+
+from collections import namedtuple
+from operator import add, or_
+
+from migen import *
+from misoc.interconnect.stream import Endpoint
+from misoc.cores.cordic import Cordic
+from artiq.gateware.rtio import rtlink
+
+
+class Dac(Module):
+    """Output module.
+
+    Holds the two output line executors.
+
+    Attributes:
+        data (Signal[16]): Output value to be send to the DAC.
+        clear (Signal): Clear accumulated phase offset when loading a new
+                        waveform. Input.
+        i (Endpoint[]): Coefficients of the output lines.
+    """
+    def __init__(self):
+        self.clear = Signal()
+        self.data = Signal(16)
+
+        ###
+
+        subs = [
+            Volt(),
+            Dds(self.clear),
+        ]
+
+        self.sync.rio += [
+            self.data.eq(reduce(add, [sub.data for sub in subs])),
+        ]
+
+        self.i = [ sub.i for sub in subs ]
+        self.submodules += subs
+
+
+class Volt(Module):
+    """DC bias spline interpolator.
+
+    The line data is interpreted as a concatenation of:
+
+        * 16 bit amplitude offset
+        * 32 bit amplitude first order derivative
+        * 48 bit amplitude second order derivative
+        * 48 bit amplitude third order derivative
+
+    Attributes:
+        data (Signal[16]): Output data from this spline.
+        i (Endpoint): Coefficients of the DC bias spline, along with its
+                        latency compensation.
+    """
+    def __init__(self):
+        self.data = Signal(16)
+        self.i = Endpoint([("data", 144)])
+
+        ###
+
+        v = [Signal(48) for i in range(4)] # amp, damp, ddamp, dddamp
+
+        # Increase latency of stb by 17 cycles to compensate CORDIC latency
+        stb_r = [ Signal() for _ in range(17) ]
+        self.sync.rio += [
+            stb_r[0].eq(self.i.stb),
+        ]
+        for idx in range(16):
+            self.sync.rio += stb_r[idx+1].eq(stb_r[idx])
```

Comment:
There should probably be a latency for the data as well, otherwise the coefficients can be wiped out before the `stb` come in effect.